[IMP] sale*: merge configuring products and opt. products modals

kzhekov opened the pull request:
The changes concern the following modules:
- sale
- sale_product_configurator
- website_sale
- website_sale_product_configurator

When optional products are enabled and a user adds an item to the cart, the
modal popup does not allow further modifications to the main product based
on its template. If the user wants to change the color or the material, they
have to close the popup, modify the product and open it back up.

The same is applied when Add to Cart is active, and the user cannot configure
the product at all after clicking the add to cart button.

This change will make it possible to change the product variants without having
to close the modal, and configure the product when using the "Add to cart"
button when it's enabled.

The added configuration step is needed for some cases like the following:

1. When adding a product that had variants directly from the /shop page before,
the first variant was added by default and configuring it was impossible without
explicitly going on the product's page and configuring it from there. This 
behaviour made no sense, and it's the reason the "optional products modal"
(which should no longer be called that) is opened when the product is not 
configured even if there are no optional products.
2. When going through the product's page, and there are optional products enabled
for the product getting configured, then it's nice but not necessary to be able to 
configure the product further while choosing and configuring the optional products.
This doesn't add an extra step in any case, it just makes it possible to still configure
your product in the modal.
3. If there are no optional products, the configuration made on the product's
page is taken into account and the product is directly added to the cart without 
opening the modal, which is the same behaviour as previously.

Only one modal is shown, depending on the situation:

- For the website flow, if the product is not configured and has variants or has
optional products (or both), a popup shows that allows for configuration + 
choosing and configuring optional products.
- For the sales flow, there was already a configuration popup, so the optional
products modal is only shown when there are optional products and replaces
the first instead.

Some tests were adapted to function with the new product configuration modal:

- The steps checking for the text mentioned above are removed
- Triggers were adapted to the modal for products that were previously added to cart
  with their default variant
- Various other small adjustments

task-2541663

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/73086).

Feyensv reviewed (commented):
Could you add the task ID to your commits and to your PR message ? :)
Also, is this task ready for review? If not, you should create your PR's in draft until they are ready for review ;).

FYI your git setup seems to be wrongly configured since github doesn't link the commit to your github account.  You should probably add the email you use as git author to your github account.

kzhekov commented:
> Could you add the task ID to your commit and to your PR message ? :)
> Also, is this task ready for review? If not, you should create your PR's in draft until they are ready for review ;).
> 
> FYI your git setup seems to be wrongly configured since github doesn't link the commit to your github account. You should probably add the email you use as git author to your github account.

Hello! I will add the task ID to both :) The task should be ready for review, although I am pretty sure there are going to be things to fix here and there, since this is my first real feature implementation.

I will look into my git configuration, it might be using my @odoo mail instead of the one linked to my github account. Thanks for the tips!

kzhekov commented on addons/sale_product_configurator/static/tests/tours/product_configurator_advanced_ui.js:
```diff
@@ -78,30 +78,6 @@ tour.register('sale_product_configurator_advanced_tour', {
-}, {
-    trigger: '.main_product div:contains("PA8: PAV1")',
-    extra_trigger: '.oe_optional_products_modal',
-    run: function () {} //check
```
These checks are no longer required, as the modal's main product description no longer contains the custom attributes previously configured. Those are carried over to the values in the newly added variant configuration options in the modal. The extra test makes sure those values are correctly carried over.

Feyensv commented on addons/sale/static/src/js/variant_mixin.js:
```diff
@@ -448,6 +483,7 @@ var VariantMixin = {
         }
         this._toggleDisable($parent, isCombinationPossible);
 
+
```
useless diff

Feyensv commented on addons/sale_product_configurator/controllers/main.py:
```diff
@@ -61,7 +61,8 @@ def _show_optional_products(self, product_id, variant_values, pricelist, handle_
         has_optional_products = product.optional_product_ids.filtered(lambda p: p._is_add_to_cart_possible(combination))
 
-        if not has_optional_products:
+        if not has_optional_products and (product.product_variant_count <= 1 or variant_values):
+            # The modal is not shown if there are no optional products and the main product either has no variants or is already configured
```
I don't really understand the change here...
Why would we show the optional products if there are no optional products ? And if we want to do so, the method name should be updated :thinking:.

JKE-be replied:
> Why would we show the optional products if there are no optional products

As you spotted, the name is not more exact...

But we open it to allow user to customize the product added in his cart from the /shop page...
I didnt check the PR and how it has been done... You are too fast for me...

But goal of this task is when you add a Customize Desk on /shop page, before we add for you the first variant of the product.template shown... now, we propose you to customize it and chose which variant you want.

How to test the functionality: add customize desk from /shop page before / after this pr

kzhekov replied:
@Feyensv I will try to find a better name for the functionality, it's indeed not very clear. 

**The added configuration step is needed for some cases like the following:**

1. When adding a product that had variants directly from the /shop page before, the first variant was added by default and configuring it was impossible without explicitly going on the product's page and configuring it from there. This behaviour made no sense.
2. When going through the product's page, and you have optional products enabled for the product you are configuring, then it's *nice* but not necessary to be able to configure the product further while choosing and configuring the optional products. This doesn't add an extra step in any case, it just makes it possible to still configure your product in the modal.
3. If there are no optional products, the configuration you have made on the product's page is taken into account and the product is directly added to the cart without opening the modal, which is the same behaviour as previously.

The bug you mention where the values you've entered are not kept when you close the modal is something that was already the case before the change, and I can look into fixing it and adding the fix to this PR if deemed necessary 😃

rdeodoo replied:
Was this taken into account?
> Why would we show the optional products if there are no optional products ? **And if we want to do so, the method name should be updated .**

Feyensv reviewed (commented):
Honestly, I'm not convinced by this feature :thinking:
What's the point of a 2 step wizard if everything can be done on the second step ?
Furthermore, If you change anything on the main product on the second step, if you try the "Back" button, you won't keep your changes on the second step.
I didn't record/remember exactly how, but I managed to reach inconsistent data by switching multiple times between first and second step of the wizard.

I posted the little comments/questions I had while reading the diff, but I should do a better tech review once the functional debate is over :D.

JKE-be commented:
@sbuhl @pierre-pde  Can you confirm this feature ? 

Since @Feyensv, suggest that we should just cancel this task...

Feyensv commented:
> Since @Feyensv, suggest that we should just cancel this task...

I'm mainly confused by the two step configurator when there are options (in sale flows, not e-commerce) and believes this can be  improved (functionally and technically) but not strictly opposed to the task :).

> I didnt check the PR and how it has been done... You are too fast for me...

If we don't review it fast, we either forget the review, or we get spammed for a lot of PR changes :D (since we unsubscribe after our review is done).
But it could be a good idea to leave PR's in draft until they passed Testing India & Testing Be to request other teams review later in the process.  I don't mind looking at PR's later in the review process, it could reduce our pings and our work ^^.

kzhekov commented:
@Feyensv I will try to find a better name for the functionality, it's indeed not very clear. 

**The added configuration step is needed for some cases like the following:**

1. When adding a product that had variants directly from the /shop page before, the first variant was added by default and configuring it was impossible without explicitly going on the product's page and configuring it from there. This behaviour made no sense, and it's the reason the "optional products modal" (which should no longer be called that) is opened when the product is not configured even if there are no optional products.
2. When going through the product's page, and you have optional products enabled for the product you are configuring, then it's *nice* but not necessary to be able to configure the product further while choosing and configuring the optional products. This doesn't add an extra step in any case, it just makes it possible to still configure your product in the modal.
3. If there are no optional products, the configuration you have made on the product's page is taken into account and the product is directly added to the cart without opening the modal, which is the same behaviour as previously.

The bug you mention where the values you've entered are not kept when you close the modal is something that was already the case before the change, and I can look into fixing it and adding the fix to this PR if deemed necessary 😃

Feyensv commented:
@kzhekov Thanks for the clarification :+1:
I now understand the logic of this change, and indeed, it seems to be a nice improvement for the e-commerce.
What I was speaking about is that your changes in the product configurator also impacts the `sale` flows.

Backend `sale` configurator flow (adding a configurable product on a sale order line, e.g. the customizable desk)
* First page, you configure the product (if the product has configurable attributes)
* Second page, you choose optional products (if any, otherwise step is skipped)
  * If you want to go back to first page, there is a "Back" button
* Third page = renting/ticketing configuration (sale_renting/event_sale) can open in some circumstances (but not for the customizable desk by default).

My point was that, in the backend `sale` flow, the new ability to modify the product attributes on the second page is strange (and doesn't work correctly):
* If you specify attributes on the first page, validate, and modifies the attribute on the second page, if you go back to first page, you loose the changes made on the second page (or part of it, as I said, I got strange results :D)
* If you can configure the attributes on the second page, what's even the point of the first page ?  We could only display the second page when there are optional products :thinking:.

I understand the e-commerce needs, but I wanted to make sure the impact(s) on the backend are considered (functionally & technically) ;).

kzhekov commented:
@Feyensv Ah yes, that makes sense, I could make it so that in all cases there is only one step for configuring the product and eventually choosing optional products from the back-end, and not two which is redundant and currently the case in some scenarios.

If that's OK with you too @sbuhl @pierre-pde ?

sbuhl commented:
@kzhekov 
Actually, is step 1 still needed in the sale flow (backend)?

kzhekov commented:
> @kzhekov
> Actually, is step 1 still needed in the sale flow (backend)?

No, in this case, the back-end would directly call the modal that contains the product to configure if there are variants and the optional products if there are any. The first step would be removed from the backend flow.

Feyensv commented:
> Actually, is step 1 still needed in the sale flow (backend)?

Note that for highly customizable products (lots of attributes to choose from), I doubt it would render as beautifully on the second page, with the optional products :thinking:

kzhekov commented:
> > Actually, is step 1 still needed in the sale flow (backend)?
> 
> Note that for highly customizable products (lots of attributes to choose from), I doubt it would render as beautifully on the second page, with the optional products

What sort of solution do you think would be best then? From my point of view, the changes in this PR improve the flow for sales as well (if I fix the current issues), and I was working on skipping the first step but if you think that it will still have a negative impact on the flow, then what would be a solution?

Feyensv commented:
> What sort of solution do you think would be best then? From my point of view, the changes in this PR improve the flow for sales as well (if I fix the current issues), and I was working on skipping the first step but if you think that it will still have a negative impact on the flow, then what would be a solution?

I don't know :D.  Ideally I would say to not change the backend flow but this would mean complexifying the product configurator logic which is already quite complex ^^.
I'll leave it to the PO's decision, it's more a functional choice.  It just has to be done considering all the impacted flows :).

kzhekov commented:
> > What sort of solution do you think would be best then? From my point of view, the changes in this PR improve the flow for sales as well (if I fix the current issues), and I was working on skipping the first step but if you think that it will still have a negative impact on the flow, then what would be a solution?
> 
> I don't know :D. Ideally I would say to not change the backend flow but this would mean complexifying the product configurator logic which is already quite complex ^^.
> I'll leave it to the PO's decision, it's more a functional choice. It just has to be done considering all the impacted flows :).

But in this case, the sale flow before was the same, except that you couldn't configure the product in the modal before. If in most cases people don't configure the product additionally in the modal, it's just a nicer way of displaying things. It didn't add any additional steps that weren't there before.

Or perhaps I don't understand what you mean by flow? But from my point of view, if the number of interactions with the UI don't change, just having the possibility to still configure your product in the optional products selection screen is positive in some cases and the same as before in most cases. Or do you have a case in mind where the user has to click on more stuff than before?

Feyensv commented:
> > Actually, is step 1 still needed in the sale flow (backend)?
> 
> Note that for highly customizable products (lots of attributes to choose from), I doubt it would render as beautifully on the second page, with the optional products

Well, I tested it again with a highly configurable product (sthg like 8 attributes with multiple variants) and it still works quite correctly.

> But in this case, the sale flow before was the same, except that you couldn't configure the product in the modal before. If in most cases people don't configure the product additionally in the modal, it's just a nicer way of displaying things. It didn't add any additional steps that weren't there before.

Honestly, I'm still not convinced that being able to modify again the product is an improvement to the sale flow.  The new display is an improvement when there isn't much attributes otherwise you have to scroll to browse the optional products, which reduces the usability of the second page.

Considering the website needs, the changes are still ok, but at least the flow when clicking on the "Back" button should be fixed (or the back button removed? :thinking:) because now, if you click on "Back" on the second page, you land on a page without even a default configuration, you have to reselect all the values before you can validate the product.

![image](https://user-images.githubusercontent.com/19364378/125934164-772b69b4-613c-436a-8b3d-057a61a528e0.png)
![image](https://user-images.githubusercontent.com/19364378/125934237-2838f2fb-990c-4fbb-b4c8-864dce0a9418.png)

kzhekov commented:
@Feyensv Could you take a look at the behaviour now and tell me what you think about it? I personally think it could be improved a bit visually, but in terms of functionality it seems to fit the needs both for the website flow and the sales flow. Do you agree?

Feyensv reviewed (commented):
> @Feyensv Could you take a look at the behaviour now and tell me what you think about it? I personally think it could be improved a bit visually, but in terms of functionality it seems to fit the needs both for the website flow and the sales flow. Do you agree?

Yep, it's better :+1: (And perfect to keep the first page when there are no optional products :tada:)

I just found a strange behavior on the grid/matrix configurator (dunno if you know this one :D), which seems caused by your dev because I cannot reproduce it on master :thinking:.

If you add a customizable product to the SO, then try to open the matrix configurator with the dedicated product ("My company tshirt (GRID)"), the matrix doesn't open, but it does if the SO is new/empty :confused:.

kzhekov commented:
> > @Feyensv Could you take a look at the behaviour now and tell me what you think about it? I personally think it could be improved a bit visually, but in terms of functionality it seems to fit the needs both for the website flow and the sales flow. Do you agree?
> 
> Yep, it's better  (And perfect to keep the first page when there are no optional products )
> 
> I just found a strange behavior on the grid/matrix configurator (dunno if you know this one :D), which seems caused by your dev because I cannot reproduce it on master .
> 
> If you add a customizable product to the SO, then try to open the matrix configurator with the dedicated product ("My company tshirt (GRID)"), the matrix doesn't open, but it does if the SO is new/empty .

Thanks for the test ! Didn't know about the matrix configuration, I'll look into the bug and what's causing it :)  I have an idea of what is causing it, so it should be a quick fix.

kzhekov commented:
> > @Feyensv Could you take a look at the behaviour now and tell me what you think about it? I personally think it could be improved a bit visually, but in terms of functionality it seems to fit the needs both for the website flow and the sales flow. Do you agree?
> 
> Yep, it's better  (And perfect to keep the first page when there are no optional products )
> 
> I just found a strange behavior on the grid/matrix configurator (dunno if you know this one :D), which seems caused by your dev because I cannot reproduce it on master .
> 
> If you add a customizable product to the SO, then try to open the matrix configurator with the dedicated product ("My company tshirt (GRID)"), the matrix doesn't open, but it does if the SO is new/empty .
@Feyensv I think i've fixed it, also added some cleanup when pressing the "Back" button. Everything should be fine with the grid configurator now.

@sbuhl @JKE-be Could you take a look and tell me if you can think of any improvements (visual, etc.) or if it's fine as it is?

rdeodoo reviewed (commented):
First superficial review
Flows are working as expected, thank you @Feyensv for the reviews

The most important stuff here is this message that explains everything : https://github.com/odoo/odoo/pull/73086#issuecomment-880532634
Just put it inside the commit message IMHO :) and your PR description, and inside the odoo task. It summarizes everything in a few sentances

rdeodoo commented on addons/sale_product_configurator/views/templates.xml:
```diff
@@ -93,22 +97,31 @@
                 </td>
                 <td class="text-center td-price" name="price">
-                    <ul class="d-none js_add_cart_variants" t-att-data-attribute_exclusions="{'exclusions: []'}"></ul>
-                    <div class="d-none oe_unchanged_value_ids" t-att-data-unchanged_value_ids="variant_values" ></div>
```
here you remove `unchanged_value_ids` data attribute and `oe_unchanged_value_ids` class, which leave dead code -> javascript using it [here](https://github.com/odoo-dev/odoo/blame/b79202a9c8cacb7d2b42ceeaafe4fcbf524a0aa2/addons/sale/static/src/js/variant_mixin.js#L289)
To check if wanted or not, is it is, delete the dead code

kzhekov replied:
Ah yes, it doesn't make sense to keep that logic in the JS file, there is no longer a need to store the original main product values since it can still get modified and the values extracted from the modal.

rdeodoo commented on addons/sale_product_configurator/views/templates.xml:
```diff
@@ -93,22 +97,31 @@
+                            <t t-set="combination" t-value="product.product_tmpl_id._get_first_possible_combination()"/>
+                        </t>
+                        <t t-if="combination" t-call="sale.variants">
+                            <t t-set="ul_class" t-value="'flex-column'" />
```
```suggestion
                            <t t-set="ul_class" t-valuef="flex-column" />
```
it is the purpose of `t-valuef`

rdeodoo commented on addons/sale_product_configurator/views/templates.xml:
```diff
@@ -32,6 +32,10 @@
 
             <input type="hidden" class="product_template_id" t-att-value="product.id"/>
             <input type="hidden" class="product_id" t-attf-name="product_id" t-att-value="product_variant.id"/>
+            <input type="hidden"
+                   class="has_optional_products"
+                   t-attf-name="has_optional_products"
+                   t-att-value="product_variant.optional_product_ids.filtered(lambda p: p._is_add_to_cart_possible(combination))"/>
```
```suggestion
            <input type="hidden" class="has_optional_products" t-attf-name="has_optional_products"
                   t-att-value="product_variant.optional_product_ids.filtered(lambda p: p._is_add_to_cart_possible(combination))"/>
```
it is nice to split line by attribute (depend on who you ask tho :yum:), but don't "over-do" it, group at least a few attribute on the same line, or at least don't do one attribute per line when your attribute is 10 char

rdeodoo commented on addons/sale_product_configurator/static/src/js/product_configurator_renderer.js:
```diff
@@ -16,12 +16,14 @@ var ProductConfiguratorFormRenderer = FormRenderer.extend(VariantMixin, {
      */
     start: function () {
         var self = this;
+        this.displayOptModal = false;
```
why do you need to duplicate this line both in the init and in the start?

rdeodoo commented on addons/sale_product_configurator/static/src/js/product_configurator_modal.js:
```diff
@@ -229,10 +236,15 @@ var OptionalProductsModal = Dialog.extend(ServicesMixin, VariantMixin, {
+            $.each(this.rootProduct.product_custom_attribute_values, function () {
+                if (this.custom_value) {
+                    let $customInput = $modalContent
+                        .find('.main_product')
+                        .find("[data-is_custom='True']")
```
```suggestion
                        .find('.main_product [data-is_custom='True']")
```
isn't that the same?

rdeodoo commented on addons/sale_product_configurator/static/src/js/product_configurator_modal.js:
```diff
@@ -180,14 +180,21 @@ var OptionalProductsModal = Dialog.extend(ServicesMixin, VariantMixin, {
         var self = this;
-        var products = [this.rootProduct];
-        this.$modal.find('.js_product.in_cart:not(.main_product)').each(function () {
+        var products = [];
```
var :)

rdeodoo commented on addons/sale_product_configurator/static/src/js/product_configurator_controller.js:
```diff
@@ -247,7 +253,8 @@ var ProductConfiguratorFormController = FormController.extend({
      */
     _onModalClose: function () {
-        if (this.renderer.state.context.configuratorMode === 'options'
+        if ((this.renderer.state.context.configuratorMode === 'options'
+            || this.renderer.state.context.configuratorMode === 'add')
```
```suggestion
        if (['options', 'add'].includes(this.renderer.state.context.configuratorMode)
```

rdeodoo commented on addons/sale_product_configurator/static/src/js/product_configurator_controller.js:
```diff
@@ -52,6 +52,12 @@ var ProductConfiguratorFormController = FormController.extend({
             this.$el.closest('.modal').addClass('d-none');
         }
+
+        if (this.renderer.displayOptModal && this.renderer.state.context.configuratorMode !== 'edit') {
+            this.$el.closest('.o_dialog_container').addClass('d-none');
+        } else {
+            this.$el.closest('.o_dialog_container').removeClass('d-none');
+        }
```
```suggestion
        this.$el.closest('.o_dialog_container').toggleClass(
            'd-none',
            this.renderer.displayOptModal && this.renderer.state.context.configuratorMode !== 'edit')
        );
```
or
```javascript
const myvar =  this.renderer.displayOptModal && this.renderer.state.context.configuratorMode !== 'edit';
this.$el.closest('.o_dialog_container').toggleClass( 'd-none', myvar);
```

To check, condition might be the otherway around, but avoid duplicating lines when you can :)

rdeodoo commented on addons/sale/static/src/js/variant_mixin.js:
```diff
@@ -565,6 +602,20 @@ var VariantMixin = {
      */
     _toggleDisable: function ($parent, isCombinationPossible) {
         $parent.toggleClass('css_not_available', !isCombinationPossible);
+        if ($parent.hasClass('in_cart')) {
+            let $modalContent = $parent.parents('.modal-content');
+            let $footer = $modalContent.find('.modal-footer');
+            let $calledFromShop = $modalContent.find('.oe_optional_products_modal.oe_website_sale');
+            let primaryButton = $footer.find('.btn-primary');
+            primaryButton.prop('disabled', !isCombinationPossible);
+            primaryButton.toggleClass('disabled', !isCombinationPossible);
+
+            if ($calledFromShop.length) {
+                let secondaryButton = $footer.find('.btn-secondary');
+                secondaryButton.prop('disabled', !isCombinationPossible);
+                secondaryButton.toggleClass('disabled', !isCombinationPossible);
+            }
+        }
```
You don't have to agree but don't you think it is easier to read when written like this?
Less use of variable, no duplicate line
```javascript
if ($parent.hasClass('in_cart')) {
    const $modalContent = $parent.parents('.modal-content');
    const btnToHandle = ['.btn-primary'];
    if ($modalContent.find('.oe_optional_products_modal.oe_website_sale').length) {
        // If called from shop
        btnToHandle.push('.btn-secondary');
    }
    $modalContent.find('.modal-footer').find(btnToHandle.join(','))
                 .prop('disabled', !isCombinationPossible)
                 .toggleClass('disabled', !isCombinationPossible);
}
```

`oe_website_sale` mentioned inside the `sale` module.. :cry: but don't worry, I saw it was already mentionned before your PR so.. keep it like that :shrug:

kzhekov replied:
I've moved it to the website_sale variant_mixin.js and override it with the website_sale specific logic. Also reduced the number of variables :D

kzhekov commented on addons/sale_product_configurator/views/templates.xml:
```diff
@@ -93,22 +97,31 @@
                 <td class='td-img'>
-                    <img t-if="product_variant" t-att-src="'/web/image/product.product/%s/image_128' % product_variant.id" alt="Product Image"/>
-                    <img t-else="" t-att-src="'/web/image/product.template/%s/image_128' % product.id" alt="Product Image"/>
+                    <t t-call="website_sale.shop_product_carousel"/>
```
Yes, I should've changed the name, this was originally intended for website_sale only but it actually belongs to the sale_product_configurator module and is used in both the sales backend flow and the website_sale shop. I'll rename those 👍🏻

kzhekov replied:
Fixed, was simply about adding a class to the image, no need for the website_sale carousel call.

rdeodoo replied:
;-)

rdeodoo commented:
@Feyensv I understood that your concern was to not add a step or part of a screen that is useless in the backend, and keep the behavior as much as possible as before when possible.

I think everything is now fine, but what I would do on top on the actual PR is to remove the backend first step (used when there is no optional product, the same screen as before this PR) and all its related JS, and just use the new screen without the optional products part.

This would:
1. Keep the same UI for end user, regardless if they have optional product or not, the popup is the same except the optional product part that is displayed or not, instead of having a different screen for products regarding if they have optional product or not -> **Better UI consistency**
2. Remove code related to that first step (XML/JS/PY), the less code the better -> **Less code to maintain**
3. The new screen without the optional product part is actually more compact, so if the product has many variant attribute, it would take more space / less scroll (which was one of your concern)

I don't even say anything about which one is the most beautiful screen as this is purely subjective (for the record I prefer the new one), but we can at least say it is not worst or better.

I guess the only remark would be that the new screen doesn't allow to change the product inside the modal, but is that even useful?

tl;dr:
When a product has no optional product, instead of keeping the first step (that existed priori to this pr):
![image](https://user-images.githubusercontent.com/30048408/128488652-98385906-259d-400a-b881-85077e0d3f0a.png)
Use the new screen without the optional products part:
![image](https://user-images.githubusercontent.com/30048408/128488775-33995d56-27b8-40a2-b2f7-f01f5320a306.png)

Feyensv commented:
> I think everything is now fine, but what I would do on top on the actual PR is to remove the backend first step (used when there is no optional product, the same screen as before this PR) and all its related JS, and just use the new screen without the optional products part.
> 
> This would:
> 
> 1. Keep the same UI for end user, regardless if they have optional product or not, the popup is the same except the optional product part that is displayed or not, instead of having a different screen for products regarding if they have optional product or not -> **Better UI consistency**

I don't mind the change, if it's accepted/validated by a PO (PDE and/or new sales PO) :).

> 2. Remove code related to that first step (XML/JS/PY), the less code the better -> **Less code to maintain**

I don't know exactly what's used only in the backend or is really shared between backend and e-commerce, but as long as you make sure everything's still working fine (rental, matrix, configurator, ...), I'm not opposed to any cleanup of the configurator logic :grinning:.

> 3. The new screen without the optional product part is actually more compact, so if the product has many variant attribute, it would take more space / less scroll (which was one of your concern)
> 
> I guess the only remark would be that the new screen doesn't allow to change the product inside the modal, but is that even useful?

I don't think so :thinking:, and I'm not even sure it works perfectly fine to change the product in the configurator :D.

pierre-pde commented:
@Feyensv  & @kzhekov , I confirm I appreciate the UI consistency, let's go!

rdeodoo reviewed (changes requested):
Commit title is wrong
only `sale_product_configurator` is mentioned but you modified more than this module.
You will probably need to shortcut the list of modules with *, also take the occasion to rephrase it so it fit 'if applied, this commit will.. [commit title]`

rdeodoo reviewed (changes requested):
Runbot is red, can you re-green the task once fixed?
Looks like a tour in sale is broken now

kzhekov commented:
> Runbot is red, can you re-green the task once fixed?
> Looks like a tour in sale is broken now

Yes, in refactoring the code and removing the first screen I realized there's a significant bug (product dynamic variants are not created when added from the new modal) and the test reflects that now that it was adapted. Hopefully I can fix it fast.

rdeodoo commented on addons/sale_product_configurator/controllers/main.py:
```diff
@@ -56,14 +56,9 @@ def _optional_product_items(self, product_id, pricelist, **kw):
         })
 
-    def _show_optional_products(self, product_id, variant_values, pricelist, handle_stock, **kw):
+    def _show_advanced_configurator(self, product_id, variant_values, pricelist, handle_stock, **kw):
```
I don't know why every controller here do:

```python
@http.route(/xxx)
def my_method(a,b,c):
    return self._my_method(...)

def _my_method(...)
    # do stuff here
```

When `_my_method()` is only called from that controller.. what's the point of that encapsulation :shrug: 

But it was like that before so.. keep it like that and ignore this message :grimacing:

rdeodoo reviewed (changes requested):
Good work anyway as product configurator is not the easiest code to work with.
Minor stuff here more on code style than anything else.
Once done, PDE might have a final test, you can ping me on task once green I will move it to his pipe

JKE-be commented:
@robodoo delegate=rdeodoo

rdeodoo commented:
@robodoo r+
Good job, this wasn't a particularly easy one.
